File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        def dfs(target, path, start):
            if target == 0:
                res.append(path[:])
                return

            for i in range(start, len(candidates)):
                
                if i > start and candidates[i] == candidates[i - 1]:
                    continue

                num = candidates[i]
                if target - num >= 0:
                    path.append(num)
                    dfs(target - num, path, i + 1)
                    path.pop()

        dfs(target, [], 0)
        return res

File: Data Structures & Algorithms/max-area-of-island/submission-2.cpp
class Solution {
public:
    int res =0;
    void dfs(int r , int c , vector<vector<int>>& grid , int& area){


        if(r >= grid.size() || c >= grid[0].size() || r < 0 || c < 0||grid[r][c] == 0){
            return;
        }
        
        area += 1;
        grid[r][c] = 0;

        dfs(r+1 , c , grid , area);
        dfs(r-1 , c , grid , area);
        dfs(r , c+1 , grid , area);
        dfs(r , c-1 , grid , area);

        res = max(area , res);




    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        for(int i = 0 ; i < grid.size() ; i++){
            for(int j = 0 ; j < grid[0].size() ; j++){
                if (grid[i][j] == 1){
                    int area = 0;
                    dfs( i , j , grid , area);
                }
                
            }
        }

        return res;
        
    }
};
